extensions/websocket/02_websocketstream.js

```diff
@@ -0,0 +1,403 @@
+      },
+      {
+        key: "signal",
+        converter: webidl.converters.AbortSignal,
+      },
+    ],
+  );
+  webidl.converters.WebSocketCloseInfo = webidl.createDictionaryConverter(
+    "WebSocketCloseInfo",
+    [
+      {
+        key: "code",
+        converter: webidl.converters["unsigned short"],
+      },
+      {
+        key: "reason",
+        converter: webidl.converters.USVString,
+        defaultValue: "",
+      },
+    ],
+  );
+
+  /**
+   * Tries to close the resource (and ignores BadResource errors).
+   * @param {number} rid
+   */
+  function tryClose(rid) {
+    try {
+      core.close(rid);
+    } catch (err) {
+      // Ignore error if the socket has already been closed.
+      if (!(err instanceof Deno.errors.BadResource)) throw err;
+    }
+  }
+
+  const _rid = Symbol("[[rid]]");
+  const _url = Symbol("[[url]]");
+  const _connection = Symbol("[[connection]]");
+  const _closed = Symbol("[[closed]]");
+  const _closing = Symbol("[[closing]]");
+  const _earlyClose = Symbol("[[earlyClose]]");
+  class WebSocketStream {
+    [_rid];
+
+    [_url];
+    get url() {
+      return this[_url];
+    }
+
+    constructor(url, options) {
+      const prefix = "Failed to construct 'WebSocketStream'";
+      webidl.requiredArguments(arguments.length, 1, { prefix });
+      url = webidl.converters.USVString(url, {
+        prefix,
+        context: "Argument 1",
+      });
+      options = webidl.converters.WebSocketStreamOptions(options, {
+        prefix,
+        context: "Argument 2",
+      });
+
+      const wsURL = new URL(url);
+
+      if (wsURL.protocol !== "ws:" && wsURL.protocol !== "wss:") {
+        throw new DOMException(
+          "Only ws & wss schemes are allowed in a WebSocket URL.",
+          "SyntaxError",
+        );
+      }
+
+      if (wsURL.hash !== "" || StringPrototypeEndsWith(wsURL.href, "#")) {
+        throw new DOMException(
+          "Fragments are not allowed in a WebSocket URL.",
+          "SyntaxError",
+        );
+      }
+
+      this[_url] = wsURL.href;
+
+      if (
+        options.protocols.length !==
+          new Set(
+            ArrayPrototypeMap(
+              options.protocols,
+              (p) => StringPrototypeToLowerCase(p),
+            ),
+          ).size
+      ) {
+        throw new DOMException(
+          "Can't supply multiple times the same protocol.",
+          "SyntaxError",
+        );
+      }
+
+      const cancelRid = core.opSync(
+        "op_ws_check_permission_and_cancel_handle",
+        this[_url],
+        true,
+      );
+
+      if (options.signal?.aborted) {
+        core.close(cancelRid);
+        const err = new DOMException(
+          "This operation was aborted",
+          "AbortError",
+        );
+        this[_connection].reject(err);
+        this[_closed].reject(err);
+      } else {
+        const abort = () => {
+          core.close(cancelRid);
+        };
+        options.signal?.addEventListener("abort", abort);
```

Comment:
Instead of using `addEventListener` use the `abortSignal.add` method to add an internal listener. The "abort" event could be dispatched manually by a user.